### PR TITLE
UCP/API: Error-handling introduced (NOT YET IMPLEMENTED)

### PR DESCRIPTION
**Pull request by alex--m**

### CODE REVIEW

**Comment by alex--m** [comment]
Goes on top of #682

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1757/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1149/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1758/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1150/ for details (Mellanox internal link).

**Comment by alex--m** [comment]
@yosefe - UCP now separated

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -796,6 +796,9 @@ void ucp_ep_destroy(ucp_ep_h ep);
  * @a ep prior to this call are completed both at the origin and at the target
  * @ref ucp_ep_h "endpont" when this call returns.
  *
+ * @note The call flushes outstanding communication operations and therefore
+ * it may return error codes associated with the operations.
```
Comment (abouteiller) [inline src/ucp/api/ucp.h]: these operations

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1689,6 +1692,8 @@ ucs_status_t ucp_worker_fence(ucp_worker_h worker);
  * "flush" and @ref ucp_worker_fence "fence" operations please see
  * @ref ucp_worker_fence "ucp_worker_fence()"
  *
+ * @note Errors originating from non-blocking calls may return from invocation.
```
Comment (abouteiller) [inline src/ucp/api/ucp.h]: Why is the comment different from the previous one?

File: src/ucp/tag/eager_snd.c [hunk 3]
```diff
@@ -205,35 +205,37 @@ static ucs_status_t ucp_tag_eager_contig_short(uct_pending_req_t *self)
         return status;
     }
 
-    ucp_request_complete(req, req->cb.send, UCS_OK);
+    ucp_request_complete(req, req->cb.send, status);
     return UCS_OK;
```
Comment (abouteiller) [inline src/ucp/tag/eager_snd.c]: Should return status here

File: src/ucp/tag/eager_snd.c [hunk 4]
```diff
@@ -205,35 +205,37 @@ static ucs_status_t ucp_tag_eager_contig_short(uct_pending_req_t *self)
         return status;
     }
```
Comment (abouteiller) [inline src/ucp/tag/eager_snd.c]: Maybe we should pass the status to ucp_request_complete: we want to complete requests that are erroneous.

File: src/ucp/tag/eager_snd.c [hunk 5]
```diff
@@ -205,35 +205,37 @@ static ucs_status_t ucp_tag_eager_contig_short(uct_pending_req_t *self)
...
 
 static ucs_status_t ucp_tag_eager_contig_bcopy_single(uct_pending_req_t *self)
 {
-    ucs_status_t status = ucp_do_am_bcopy_single(self, UCP_AM_ID_EAGER_ONLY,
-                                                 ucp_tag_pack_eager_only_contig);
+    ucs_status_t status;
+    status = ucp_do_am_bcopy_single(self, UCP_AM_ID_EAGER_ONLY,
+                                    ucp_tag_pack_eager_only_contig);
     if (status == UCS_OK) {
         ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
```
Comment (abouteiller) [inline src/ucp/tag/eager_snd.c]: same as above

File: src/ucp/tag/eager_snd.c [hunk 6]
```diff
@@ -205,35 +205,37 @@ static ucs_status_t ucp_tag_eager_contig_short(uct_pending_req_t *self)
...
+                                   UCP_AM_ID_EAGER_FIRST,
+                                   UCP_AM_ID_EAGER_MIDDLE,
+                                   UCP_AM_ID_EAGER_LAST,
+                                   sizeof(ucp_eager_first_hdr_t),
+                                   sizeof(ucp_eager_hdr_t),
+                                   ucp_tag_pack_eager_first_contig,
+                                   ucp_tag_pack_eager_middle_contig,
+                                   ucp_tag_pack_eager_last_contig);
     if (status == UCS_OK) {
         ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
```
Comment (abouteiller) [inline src/ucp/tag/eager_snd.c]: same as above

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1193/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1800/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1194/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1801/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1200/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1227/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1836/ for details.

**Comment by MattBBaker** [comment]
ORNL failure was an out of memory error in /dev/shm. And that directory is clean these days.
```
[ RUN      ] mm/uct_p2p_rma_test.get_bcopy/0
[     INFO ] 1 3.9g [1461763139.718301] [pod9:4213 :0]       mm_posix.c:79   UCX  ERROR Not enough memory to write total of 4208440835 bytes. Please check that /dev/shm or the directory you specified has more available memory.
[1461763140.247323] [pod9:4213 :0]        uct_mem.c:89   UCX  ERROR failed to allocate 4208440835 bytes using pd posix: Out of memory
<https://pod9.ccs.ornl.gov/job/ornl_pod9/ws/contrib/../test/gtest/uct/uct_test.cc>:215: Failure
Error: Out of memory

[  FAILED  ] mm/uct_p2p_rma_test.get_bcopy/0, where GetParam() = mm/posix (12032 ms)
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1847/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1237/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1848/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1849/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1850/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1851/ for details.

**Comment by yosefe** [comment]
will be replaced by new PR.

**Pull request closed without merging**